File: algorithms/503.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int> &nums) {
        vector<int> res(nums.size(), -1);
        stack<int> s;
        for (int i = nums.size()-1; i >= 0; i--)
            s.push(nums[i]);
        for (int i = nums.size()-1; i >= 0; i--) {
            while (!s.empty() && s.top() <= nums[i])
                s.pop();
            if (!s.empty())
                res[i] = s.top();
            s.push(nums[i]);
        }
        return res;
    }
};

int main() {
    return 0;
}

File: algorithms/python3/503.py
class Solution:
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums = nums + nums
        rls = [-1] * len(nums)
        stack = []
        for i in range(len(nums)-1, -1, -1):
            num = nums[i]
            while len(stack) and num >= stack[-1]:
                stack.pop()
            if len(stack):
                rls[i] = stack[-1]
            stack.append(num)
        return rls[:len(nums)//2]
